File: frontend/src/types/task.ts
export const IMPORTANCE_LEVELS = ['HIGH', 'MEDIUM', 'LOW'] as const
export const URGENCY_LEVELS = ['HIGH', 'MEDIUM', 'LOW'] as const

export type Importance = (typeof IMPORTANCE_LEVELS)[number]
export type Urgency = (typeof URGENCY_LEVELS)[number]
export type TaskType = 'WORK' | 'PERSONAL'

export interface Task {
  id: number
  title: string
  description: string
  deadline: string | null
  estimatedMinutes: number
  importance: Importance
  urgency: Urgency
  type: TaskType
  completed: boolean
  priorityScore: number
}

export interface CreateTaskInput {
  title: string
  description: string
  deadline: string | null
  estimatedMinutes: number
  importance: Importance
  urgency: Urgency
  type: TaskType
}

export interface PlanningSlot {
  start: string
  end: string
  type: TaskType
  taskId: number | null
  taskTitle: string | null
  allocatedMinutes: number | null
}

export interface PlanningResponse {
  slots: PlanningSlot[]
}

File: backend/src/main/java/fr/tools/timemanager/domain/model/Task.java
package fr.tools.timemanager.domain.model;

import java.time.Duration;
import java.time.LocalDateTime;

public record Task(
        Long id,
        String title,
        String description,
        LocalDateTime deadline,
        Duration estimatedTime,
        Importance importance,
        Urgency urgency,
        TaskType type,
        boolean completed
) {
    public Task {
        if (title == null || title.isBlank()) {
            throw new IllegalArgumentException("title must not be blank");
        }
        description = description == null ? "" : description.trim();
        estimatedTime = estimatedTime == null || estimatedTime.isNegative() || estimatedTime.isZero()
                ? Duration.ofMinutes(30)
                : estimatedTime;
        importance = importance == null ? Importance.MEDIUM : importance;
        urgency = urgency == null ? Urgency.MEDIUM : urgency;
        type = type == null ? TaskType.WORK : type;
    }

    public int priorityScore() {
        return importance.weight() * urgency.weight();
    }
}
